crates/swc_ecma_minifier/src/compress/optimize/params.rs

```diff
@@ -0,0 +1,363 @@
+        }
+    }
+
+    /// Apply parameter inlining transformation to a function.
+    fn apply_param_inlining(
+        &mut self,
+        f: &mut Function,
+        params_to_inline: &[(usize, Box<Expr>)],
+        fn_id: &Id,
+    ) {
+        use rustc_hash::FxHashSet;
+
+        // Sort in reverse order to remove from the end first
+        let mut sorted_params = params_to_inline.to_vec();
+        sorted_params.sort_by(|a, b| b.0.cmp(&a.0));
+
+        // Collect const declarations to add at the beginning of function body
+        let mut const_decls = Vec::new();
+
+        // Track which parameter indices are being inlined
+        let mut inlined_indices = FxHashSet::default();
+
+        for (param_idx, value) in &sorted_params {
+            inlined_indices.insert(*param_idx);
+
+            if let Some(param) = f.params.get(*param_idx) {
+                if let Pat::Ident(ident) = &param.pat {
+                    // Create const declaration: const paramName = value;
+                    const_decls.push(Stmt::Decl(Decl::Var(Box::new(VarDecl {
+                        span: DUMMY_SP,
+                        ctxt: Default::default(),
+                        kind: VarDeclKind::Const,
+                        declare: false,
+                        decls: vec![VarDeclarator {
+                            span: DUMMY_SP,
+                            name: Pat::Ident(ident.clone()),
+                            init: Some(value.clone()),
+                            definite: false,
+                        }],
+                    }))));
+                }
+            }
+        }
+
+        // Remove parameters (in reverse order)
+        for (param_idx, _) in &sorted_params {
+            f.params.remove(*param_idx);
+        }
+
+        // Add const declarations to function body
+        if let Some(body) = &mut f.body {
+            const_decls.reverse(); // Reverse to maintain original order
+            const_decls.append(&mut body.stmts);
+            body.stmts = const_decls;
+        }
+
+        // Store the inlined parameter indices for later use when visiting call
+        // sites
+        self.inlined_params.insert(fn_id.clone(), inlined_indices);
+
+        self.changed = true;
+        report_change!(
+            "params: Inlined {} parameter(s) for function '{}{:?}'",
+            params_to_inline.len(),
+            fn_id.0,
+            fn_id.1
+        );
+    }
+
+    /// Check if an expression is a safe constant value for parameter inlining.
+    fn is_safe_constant_for_param_inline(expr: &Expr) -> bool {
+        match expr {
+            // Literal values are always safe
+            Expr::Lit(Lit::Null(_))
+            | Expr::Lit(Lit::Bool(_))
+            | Expr::Lit(Lit::Num(_))
+            | Expr::Lit(Lit::Str(_))
+            | Expr::Lit(Lit::BigInt(_)) => true,
+
+            // undefined identifier (unresolved)
+            Expr::Ident(id) if id.sym == "undefined" => true,
```

Comment:
shouldn't we allow arbitrary identifiers?

```
function a(b) {...}

const x = ....;

a(x)
```

we should be able to inline `x` (i guess as long as the name is not shadowed?)

Reply:
https://github.com/swc-project/swc/actions/runs/18773516269/job/53562955301?pr=11156
AI tried it, but the CI failed in a bad way, so AI [reverted](https://github.com/swc-project/swc/pull/11156/commits/96d0c556c5e5602e24af31ee0f6fb116e53ce3dd) it.